diff --git a/node_modules/@whatwg-node/fetch/dist/index.d.ts b/node_modules/@whatwg-node/fetch/dist/index.d.ts
index 1f3d361..4286a02 100644
--- a/node_modules/@whatwg-node/fetch/dist/index.d.ts
+++ b/node_modules/@whatwg-node/fetch/dist/index.d.ts
@@ -1,37 +1,21 @@
 /// <reference lib="dom" />
 
-declare const _fetch: typeof fetch;
-declare const _Request: typeof Request;
-declare const _Response: typeof Response;
-declare const _Headers: typeof Headers;
-declare const _FormData: typeof FormData;
-declare const _AbortController: typeof AbortController;
-declare const _ReadableStream: typeof ReadableStream;
-declare const _WritableStream: typeof WritableStream;
-declare const _TransformStream: typeof TransformStream;
-declare const _Blob: typeof Blob;
-declare const _File: typeof File;
-declare const _crypto: typeof crypto;
-declare const _btoa: typeof btoa;
-declare const _TextEncoder: typeof TextEncoder;
-declare const _TextDecoder: typeof TextDecoder;
-
 declare module "@whatwg-node/fetch" {
-  export const fetch: typeof _fetch;
-  export const Request: typeof _Request;
-  export const Response: typeof _Response;
-  export const Headers: typeof _Headers;
-  export const FormData: typeof _FormData;
-  export const AbortController: typeof _AbortController;
-  export const ReadableStream: typeof _ReadableStream;
-  export const WritableStream: typeof _WritableStream;
-  export const TransformStream: typeof _TransformStream;
-  export const Blob: typeof _Blob;
-  export const File: typeof _File;
-  export const crypto: typeof _crypto;
-  export const btoa: typeof _btoa;
-  export const TextDecoder: typeof _TextDecoder;
-  export const TextEncoder: typeof _TextEncoder;
+  export const fetch: typeof globalThis.fetch;
+  export const Request: typeof globalThis.Request;
+  export const Response: typeof globalThis.Response;
+  export const Headers: typeof globalThis.Headers;
+  export const FormData: typeof globalThis.FormData;
+  export const AbortController: typeof globalThis.AbortController;
+  export const ReadableStream: typeof globalThis.ReadableStream;
+  export const WritableStream: typeof globalThis.WritableStream;
+  export const TransformStream: typeof globalThis.TransformStream;
+  export const Blob: typeof globalThis.Blob;
+  export const File: typeof globalThis.File;
+  export const crypto: typeof globalThis.crypto;
+  export const btoa: typeof globalThis.btoa;
+  export const TextDecoder: typeof globalThis.TextDecoder;
+  export const TextEncoder: typeof globalThis.TextEncoder;
   export interface FormDataLimits {
     /* Max field name size (in bytes). Default: 100. */
     fieldNameSize?: number;
@@ -49,21 +33,21 @@ declare module "@whatwg-node/fetch" {
     headerSize?: number;
   }
   export const createFetch: (opts?: { useNodeFetch?: boolean; formDataLimits?: FormDataLimits }) => ({
-    fetch: typeof _fetch,
-    Request: typeof _Request,
-    Response: typeof _Response,
-    Headers: typeof _Headers,
-    FormData: typeof _FormData,
-    AbortController: typeof _AbortController,
-    ReadableStream: typeof _ReadableStream,
-    WritableStream: typeof _WritableStream,
-    TransformStream: typeof _TransformStream,
-    Blob: typeof _Blob,
-    File: typeof _File,
-    crypto: typeof _crypto,
-    btoa: typeof _btoa,
-    TextEncoder: typeof _TextEncoder,
-    TextDecoder: typeof _TextDecoder
+    fetch: typeof globalThis.fetch,
+    Request: typeof globalThis.Request,
+    Response: typeof globalThis.Response,
+    Headers: typeof globalThis.Headers,
+    FormData: typeof globalThis.FormData,
+    AbortController: typeof globalThis.AbortController,
+    ReadableStream: typeof globalThis.ReadableStream,
+    WritableStream: typeof globalThis.WritableStream,
+    TransformStream: typeof globalThis.TransformStream,
+    Blob: typeof globalThis.Blob,
+    File: typeof globalThis.File,
+    crypto: typeof globalThis.crypto,
+    btoa: typeof globalThis.btoa,
+    TextEncoder: typeof globalThis.TextEncoder,
+    TextDecoder: typeof globalThis.TextDecoder
   });
 }
 
