diff --git a/packages/keystone/bin/commands/dev.js b/packages/keystone/bin/commands/dev.js
index 24f568b..84f894c 100644
--- a/packages/keystone/bin/commands/dev.js
+++ b/packages/keystone/bin/commands/dev.js
@@ -8,6 +8,7 @@ module.exports = {
     '--port':       Number,
     '-p':           '--port',
     '--entry':      String,
+    '--app_url':    String,
     '--connect-to': String,
   },
   help: ({ exeName }) => `
@@ -16,6 +17,7 @@ module.exports = {

     Options
       --port, -p    Port to start on [${DEFAULT_PORT}]
+      --app_url     Custom application URL
       --entry       Entry file exporting keystone instance [${DEFAULT_ENTRY}]
   `,
   exec: async (args, { exeName, _cwd = process.cwd() } = {}, spinner) => {
diff --git a/packages/keystone/bin/commands/start.js b/packages/keystone/bin/commands/start.js
index 0e8d02a..e42fcce 100644
--- a/packages/keystone/bin/commands/start.js
+++ b/packages/keystone/bin/commands/start.js
@@ -8,6 +8,7 @@ module.exports = {
     '--port':       Number,
     '-p':           '--port',
     '--entry':      String,
+    '--app_url':    String,
     '--out':        String,
     '-o':           '--out',
   },
@@ -17,6 +18,7 @@ module.exports = {

     Options
       --port, -p    Port to start on [${DEFAULT_PORT}]
+      --app_url     Custom application URL
       --entry       Entry file exporting keystone instance [${DEFAULT_ENTRY}]
   `,
   exec: async (args, { exeName, _cwd = process.cwd() } = {}, spinner) => {
diff --git a/packages/keystone/bin/utils.js b/packages/keystone/bin/utils.js
index 2f2e555..eb95ec0 100644
--- a/packages/keystone/bin/utils.js
+++ b/packages/keystone/bin/utils.js
@@ -5,7 +5,7 @@ const ciInfo = require('ci-info');
 const chalk = require('chalk');
 const path = require('path');

-const { DEFAULT_ENTRY, DEFAULT_PORT } = require('../constants');
+const { DEFAULT_ENTRY, DEFAULT_PORT, DEFAULT_APP_URL } = require('../constants');

 const verifyTableMessages = verifyTableResults => {
   const verifyTableErrors = verifyTableResults.filter(({ isRejected }) => isRejected);
@@ -45,11 +45,11 @@ const verifyTableMessages = verifyTableResults => {
   }
 };

-const ttyLink = (text, path, port) => {
+const ttyLink = (text, path, appUrl) => {
   if (ciInfo.isCI) {
     return;
   }
-  const url = `http://localhost:${port}${path}`;
+  const url = `${appUrl}${path}`;
   const link = terminalLink(url, url, { fallback: () => url });
   console.log(`🔗 ${chalk.green(text)}\t${link}`);
 };
@@ -96,6 +96,7 @@ function extractAppMeta(apps, dev) {

 async function executeDefaultServer(args, entryFile, distDir, spinner) {
   const port = args['--port'] ? args['--port'] : DEFAULT_PORT;
+  const appUrl = args['--app_url'] ? args['--app_url'] : DEFAULT_APP_URL;
   let status = 'start-server';

   spinner.text = 'Starting Keystone server';
@@ -133,7 +134,7 @@ async function executeDefaultServer(args, entryFile, distDir, spinner) {
   const {
     keystone,
     apps = [],
-    configureExpress = () => {},
+    configureExpress = () => { },
     cors,
     pinoOptions,
   } = require(path.resolve(entryFile));
@@ -163,13 +164,12 @@ async function executeDefaultServer(args, entryFile, distDir, spinner) {
   app.use(middlewares);
   status = 'started';
   spinner.succeed(chalk.green.bold(`Keystone instance is ready at http://localhost:${port} 🚀`));
-
   const { adminPath, graphiqlPath, apiPath } = extractAppMeta(apps, dev);

   /* eslint-disable no-unused-expressions */
-  adminPath && ttyLink('Keystone Admin UI:', adminPath, port);
-  graphiqlPath && ttyLink('GraphQL Playground:', graphiqlPath, port);
-  apiPath && ttyLink('GraphQL API:\t', apiPath, port);
+  adminPath && ttyLink('Keystone Admin UI:', adminPath, appUrl);
+  graphiqlPath && ttyLink('GraphQL Playground:', graphiqlPath, appUrl);
+  apiPath && ttyLink('GraphQL API:\t', apiPath, appUrl);
   /* eslint-enable no-unused-expressions */

   return { port, server };
diff --git a/packages/keystone/constants.js b/packages/keystone/constants.js
index 87a280c..fb58ff6 100644
--- a/packages/keystone/constants.js
+++ b/packages/keystone/constants.js
@@ -1,5 +1,7 @@
+const PORT = process.env.PORT || 3000;
 module.exports = {
-  DEFAULT_PORT: process.env.PORT || 3000,
+  DEFAULT_PORT: PORT,
+  DEFAULT_APP_URL: process.env.APP_URL || `http://localhost:${PORT}`,
   DEFAULT_ENTRY: 'index.js',
   DEFAULT_SERVER: 'server.js',
   DEFAULT_DIST_DIR: 'dist',
